### PR TITLE
40 : Add keycloak authentication in jpa server starter

### PR DESCRIPTION
**Pull request by rehammuzzamil**
https://github.com/OpenSRP/fhircore/issues/40

To build:

1. Update keycloak details from the stage Keycloak server in application.yml
2. Add `http://localhost:8080/*` as a redirect URL on the stage Keycloak server for this realm and client
3. Build `hapi-fhir-opensrp-security-config` package from the `hapi-fhir` repo (Reference [PR](https://github.com/opensrp/hapi-fhir/pull/1))
4. Build and run `jpa-server-starter` project using `Pboot` profile. Deploy on Tomcat server.
5. Visit http://localhost:8080/fhir/rest/Patient/1 on your browser. This should currently redirect to the Keycloak login page. The valid response is returned with a status code of 200 when the correct login credentials are entered.
6. Visit http://localhost:8080/fhir/rest/Patient/1 on Postman with the correct Oauth token to fetch the response. 

Please note : Authentication in test-page-overlay (web UI) is not a part of this PR.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -127,6 +127,12 @@
             <classifier>classes</classifier>
         </dependency>
 
+        <dependency>
+            <groupId>ca.uhn.hapi.fhir</groupId>
+            <artifactId>hapi-fhir-opensrp-security-config</artifactId>
+            <version>5.4.0-PRE5-SNAPSHOT</version>
```
Comment (ekigamba) [inline pom.xml:133]: The version for this should be start from `0.0.1`

File: src/main/resources/application.yaml [hunk 2]
```diff
@@ -1,10 +1,12 @@
...
   datasource:
-    url: 'jdbc:h2:file:./target/database/h2'
+    url: 'jdbc:postgresql://localhost:5432/hapi_fhir'
     #url: jdbc:h2:mem:test_mem
-    username: sa
-    password: null
-    driverClassName: org.h2.Driver
+    username: postgres
+    password: root
+    driverClassName: org.postgresql.Driver
```
Comment (ekigamba) [inline src/main/resources/application.yaml:9]: The secrets in this should be moved to a private repository or secrets file. @dubdabasoduba will liase with SRE and decided how to move forward with this

File: src/main/resources/application.yaml [hunk 3]
```diff
@@ -153,3 +155,15 @@ hapi:
...
 #  schema_management_strategy: CREATE
 #  username: SomeUsername
+#security:
+#  ignored: none
+keycloak:
+  auth-server-url: http://localhost:8180/auth/
+  realm: fhir-core
+  resource: fhir-core-server
+  credentials:
+    secret: b7747a19-f72f-4906-8892-8438ce2492be
```
Comment (ekigamba) [inline src/main/resources/application.yaml:165]: These too should be moved to secrets file. @rehammuzzamil Are these credentials for a local keycloak instance?

File: src/main/java/ca/uhn/fhir/jpa/starter/Application.java [hunk 4]
```diff
@@ -52,7 +55,7 @@ public ServletRegistrationBean hapiServletRegistration() {
     JpaRestfulServer jpaRestfulServer = new JpaRestfulServer();
     beanFactory.autowireBean(jpaRestfulServer);
     servletRegistrationBean.setServlet(jpaRestfulServer);
-    servletRegistrationBean.addUrlMappings("/fhir/*");
+    servletRegistrationBean.addUrlMappings("/fhir/rest/*");
```
Comment (ekigamba) [inline src/main/java/ca/uhn/fhir/jpa/starter/Application.java:58]: I am not sure why this change was made. Could you help me out?

**Review by ekigamba — CHANGES_REQUESTED** [review]

File: src/main/resources/application.yaml (hunk 3 above)
Comment (rehammuzzamil) [reply src/main/resources/application.yaml:165]: @ekigamba  yes these creds are for local keycloak instance.

Comment (ekigamba) [reply src/main/resources/application.yaml:165]: Cool @rehammuzzamil , nice work

File: src/main/java/ca/uhn/fhir/jpa/starter/Application.java (hunk 4 above)
Comment (rehammuzzamil) [reply src/main/java/ca/uhn/fhir/jpa/starter/Application.java:58]: @ekigamba this change was made to make the test page overlay work. 
As context for web UI and API both were same starting with `/fhir` therefore we have added `/fhir/rest `to distinguish the context for the APIs. With this approach, we have disabled the web security on /fhir. 

But test page overlay still does not support Authentication and is broken when we try to hit API from there. We may think to remove it as it was before and fix the test page overlay in some other issue. 

Let me re-test it after removing the `/rest` from the context.

cc: @maimoonak  @dubdabasoduba

Comment (ekigamba) [reply src/main/java/ca/uhn/fhir/jpa/starter/Application.java:58]: Cool

**Comment by ekigamba** [comment]
I have tested this locally with the stage Keycloak instance and it works as described on the PR description :+1: . The overlay page is not working as described

File: src/main/java/ca/uhn/fhir/jpa/starter/Application.java (hunk 4 above)
Comment (rehammuzzamil) [reply src/main/java/ca/uhn/fhir/jpa/starter/Application.java]: Updated back to the original one. @ekigamba 
cc : @maimoonak

File: pom.xml (hunk 1 above)
Comment (rehammuzzamil) [reply pom.xml]: Done, here is the new [repo](https://github.com/opensrp/hapi-fhir-keycloak/blob/master/pom.xml).

File: src/main/resources/application.yaml (hunk 2 above)
Comment (dubdabasoduba) [reply src/main/resources/application.yaml:9]: @manutarus @bennsimon Do you have an idea of how to use secret files to provide these creds?

File: src/main/resources/application.yaml (hunk 3 above)
Comment (dubdabasoduba) [reply src/main/resources/application.yaml:165]: @rehammuzzamil can we load this from a properties file or something. 

@bennsimon @manutarus does defining these in a separate file help in encrypting them?

**Review by dubdabasoduba — CHANGES_REQUESTED** [review]

File: src/main/resources/application.yaml (hunk 3 above)
Comment (rehammuzzamil) [reply src/main/resources/application.yaml:165]: @dubdabasoduba  these keycloak configurations need to be defined in a properties file either `application.yml` or `application.properties`. In our case, we have an` _application.yml file_`

**Review by ekigamba — APPROVED** [review]

**Review by dubdabasoduba — APPROVED** [review]

**Comment by rehammuzzamil** [comment]
As discussed, we can merge this PR for now. Discussion for encrypting secrets will be done afterward.
